### PR TITLE
Enable pylint undefined-loop-variable warning

### PR DESCRIPTION
**Pull request by uzzell**
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).

Enables the Pylint warning `undefined-loop-variable`.  This contributes to (now-closed) #48855 and extends #49740.

Replaces several instances of
```
for x in my_list:
  foo(x)
bar(x)  # undefined-loop-variable
```
with
```
last_element = None
for x in my_list:
  last_element = x
  foo(x)
bar(last_element)  # no undefined-loop-variable
```

### CODE REVIEW

File: pandas/io/excel/_base.py [hunk 1]
```diff
@@ -726,7 +726,9 @@ def parse(
 
         output = {}
 
+        last_sheetname = None
```
Comment (mroeschke) [inline pandas/io/excel/_base.py:729]: Might be good to check after the loop if `last_sheetname = None` then raise an error that `sheets` was an empty list?

Comment (uzzell) [reply pandas/io/excel/_base.py:729]: Good idea.  What error type would be most appropriate?  My thought is a ValueError, but I wonder whether something else might be better.

Comment (mroeschke) [reply pandas/io/excel/_base.py:729]: Yeah ValueError would be good

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @uzzell

**Pull request merged**